### DISCUSSION TITLE
ROOT TMVA model evaluation in Coffea-based framework

### DISCUSSION THREAD

**Discussion by alesaggio** [scikit-hep/coffea#520] [Ideas]
Hi, I would need help with a problem regarding the evaluation of a `.weights.xml` model derived from ROOT TMVA in a Coffea-based framework.
I have a model file derived from a TMVA training that provides the value of a lepton ID, given some lepton variables, when being evaluated over. The problem is that the TMVA functions (such as `TMVA.AddVariable(..)` or `TMVA.EvaluateMVA(..)`) operate lepton by lepton, which is not good for my setup because I am working with awkward arrays, and so what I need is to evaluate the model on the whole lepton array "at once". Looping over the leptons is of course not a viable solution… I believe that a way out might be to convert the TMVA model into some coffea/awkward-compatible one, so that I could evaluate my model directly on Jagged Arrays or even dataframes. Do you know what the best way to approach this is?

**Comment by lgray** [comment]
@alesaggio I would first try to see if the TMVA model can be converted into [xgboost](https://github.com/dmlc/xgboost).
From there working with awkward arrays is a snap.

There is a conversion function from xgboost to TMVA in the ROOT pacakge: 
```python3
#in pyroot
ROOT.TMVA.SaveXGBoost(xgb, "myBDT", "model.root")
```
so the opposite must be possible!

However, unsurprisingly there does not appear to be a pre-written implementation for this.

Adding:
@nsmith- @kpedro88 @jpata @bendavid
as they may, and probably do, have more info.

**Reply by kpedro88** [reply]
https://github.com/kpedro88/koza4ok (linking to my fork because it includes an important patch not merged in the upstream repo) can convert from sklearn to TMVA format. In principle, the reverse process could be implemented based on this.

**Comment by alesaggio** [comment]
Thank you for your replies. I will keep looking for options, though what you mention seems like a useful starting point.

**Comment by nsmith-** [comment]
Check out https://github.com/guitargeek/tmva-to-xgboost
